src/librustc_trans/back/write.rs

```diff
@@ -1033,94 +956,202 @@ fn execute_work_item(cgcx: &CodegenContext,
+                match diag.code {
+                    Some(ref code) => {
+                        handler.emit_with_code(&MultiSpan::new(),
+                                               &diag.msg,
+                                               &code,
+                                               diag.lvl);
+                    }
+                    None => {
+                        handler.emit(&MultiSpan::new(),
+                                     &diag.msg,
+                                     diag.lvl);
                     }
-                    None => break,
                 }
             }
-
-            tx.take().unwrap().send(()).unwrap();
-        }).unwrap();
+            Message::InlineAsmError(cookie, msg) => {
+                match Mark::from_u32(cookie).expn_info() {
+                    Some(ei) => sess.span_err(ei.call_site, &msg),
+                    None     => sess.err(&msg),
+                }
+            }
+            Message::AbortIfErrors => sess.diagnostic().abort_if_errors(),
+        }
     }
 
-    let mut panicked = false;
-    for rx in futures {
-        match rx.recv() {
-            Ok(()) => {},
-            Err(_) => {
-                panicked = true;
-            },
+    // Just in case, check this on the way out.
+    sess.diagnostic().abort_if_errors();
+}
+
+struct SharedEmitter {
+    tx: Sender<Message>,
+}
+
+impl Emitter for SharedEmitter {
+    fn emit(&mut self, db: &DiagnosticBuilder) {
+        drop(self.tx.send(Message::Diagnostic(Diagnostic {
+            msg: db.message(),
+            code: db.code.clone(),
+            lvl: db.level,
+        })));
+        for child in &db.children {
+            drop(self.tx.send(Message::Diagnostic(Diagnostic {
+                msg: child.message(),
+                code: None,
+                lvl: child.level,
+            })));
         }
-        // Display any new diagnostics.
-        diag_emitter.dump(sess.diagnostic());
-    }
-    if panicked {
-        sess.fatal("aborting due to worker thread panic");
+        drop(self.tx.send(Message::AbortIfErrors));
     }
 }
 
+fn spawn_work<'a>(sess: &'a Session,
+                  exported_symbols: &'a ExportedSymbols,
+                  scope: &Scope<'a>,
+                  tx: Sender<Message>,
+                  work: WorkItem,
+                  idx: usize) {
+    let plugin_passes = sess.plugin_llvm_passes.borrow().clone();
+    let remark = sess.opts.cg.remark.clone();
+    let incr_comp_session_dir = sess.incr_comp_session_dir_opt().map(|r| r.clone());
+    let depth = time_depth();
+    let lto = sess.lto();
+    let crate_types = sess.crate_types.borrow().clone();
+    let mut each_linked_rlib = Vec::new();
+    drop(link::each_linked_rlib(sess, &mut |cnum, path| {
+        // `#![no_builtins]` crates don't participate in LTO.
+        if sess.cstore.is_no_builtins(cnum) {
+            return
+        }
+        each_linked_rlib.push((cnum, path.to_path_buf()));
```

Comment:
If the `each_linked_rlib` field is LTO-specific, we should probably change to the name to reflect this.